### ISSUE TITLE
sql: internal error: columns required of root must be subset of output columns

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#40592]
```
CREATE TABLE IF NOT EXISTS tab_orig AS
	SELECT
		g AS _int2,
		g AS _int8,
		'2001-01-01'::TIMESTAMPTZ + g * '1 day'
			AS _timestamptz,
		g % 2 = 0 AS _bool,
		g AS _decimal,
		g AS _string
	FROM
		generate_series(0, 0) AS g;

SELECT
	NULL
FROM
	tab_orig AS tab_10387
WHERE
	EXISTS(
		SELECT
			NULL
		FROM
			tab_orig
			JOIN tab_orig AS tab_10390
				INNER JOIN tab_orig AS tab_10391
					JOIN tab_orig AS tab_10392 ON true ON
						EXISTS(
							WITH
								with_2326 AS (SELECT NULL)
							SELECT
								tab_10387._string
						) ON true
			JOIN tab_orig AS tab_10401 ON
					tab_10392._timestamptz
					= tab_10401._timestamptz
					AND tab_10392._int2 = tab_10401._int8
			JOIN tab_orig AS tab_10402 ON
					tab_10390._decimal = tab_10402._decimal
		WHERE
			tab_10391._bool
		ORDER BY
			tab_10391._decimal
		LIMIT
			66
	);
```

```
stack trace:
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:790: optimizeRootWithProps()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:203: Optimize()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/apply_join.go:259: Next()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:171: Next()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171: Run()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793: Run()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:584: Run()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:337: Run()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:962: PlanAndRun()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:816: execWithDistSQLEngine()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:708: dispatchToExecutionEngine()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:417: execStmtInOpenState()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:98: execStmt()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1238: execCmd()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1167: run()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:444: ServeConn()
/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:584: func1()
runtime/asm_amd64.s:1337: goexit()
```

**Comment by yuzefovich** [comment]
I didn't look into this, but the stack trace suggests that this is another apply join bug, and the problem seems to be somewhere in the optimizer. cc @rytaft

**Comment by rytaft** [comment]
Yea this looks like it's on us - I'll take a look - thanks!

**Comment by rytaft** [comment]
I'm not quite sure how, but this seems to have been fixed by #40673. I tried tweaking the query so that it referenced the `WithExpr` multiple times to prevent the `InlineWith` rule from triggering, but I couldn't reproduce the error. Marking this as fixed. 

cc @justinj in case you have an idea about how your change fixed this bug!
